src/Castle.Windsor/MicroKernel/Registration/ComponentDependencyRegistrationExtensions.cs

```diff
@@ -57,7 +57,7 @@ public static IDictionary InsertAnonymous(this IDictionary arguments, object nam
 		/// <summary>
 		///   Inserts a new typed argument with given type. If an argument for this type already exists, it will be overwritten.
 		/// </summary>
-		public static IDictionary InsertTyped<TDependencyType>(this IDictionary arguments, TDependencyType value)
+		public static T InsertTyped<T, TDependencyType>(this T arguments, TDependencyType value) where T: IDictionary
```

Comment:
I wonder if these extension methods should be amended to come in line with the other changes. These extension methods used `IDictionary` but because it has methods for typed arguments it sort of makes the assumption they'll be used for an `Arguments` instance, so why not just mark them on `Arguments` rather than for all `IDictionary`s which would pop up in places that are unwanted.

Do we even needed these extension methods? Are they used for something else which I've overlooked? If we keep them we'll want to bring them in line with some of the naming we've changed to with the Arguments factory methods and add `IReadOnlyDictionary`.

Reply:
Hmmm. As far as I can tell these methods are only used in the tests. It all seems a bit surplus to requirements now that we have the factory methods on Arguments. I think we can safely delete these methods. I will make another note in the CHANGELOG.md.

Reply:
You are right, we should probably move this over to Arguments, trying to get rid of them reveals that `InsertTyped` is used in [DelegateFactory](https://github.com/castleproject/Windsor/blob/master/src/Castle.Windsor/Facilities/TypedFactory/Internal/DelegateFactory.cs#L61) in typed factories. 

Are you also OK with me fixing some of the IDictionary parameters I found on delegates that appear to be used as part of a Resolve? [DynamicParametersResolveDelegate](https://github.com/castleproject/Windsor/blob/master/src/Castle.Windsor/MicroKernel/Registration/DynamicParametersResolveDelegate.cs) is an example.

Reply:
I was thinking, possibly getting rid of the delegates and using vanilla Func's. Also instead of passing IDictionary we could replace that with Arguments. That would for example allow me to move an extension like `InsertTyped` over. What do you think? I will do this as a single commit which can be reviewed separately and we can rebase out if we think it touches on too many things.

Reply:
I eventually opted for not removing the delegates. They ended up being nicer to work with and read.

Please this commit for review: https://github.com/castleproject/Windsor/pull/420/commits/6329505df9914a1a297f55de00911a3771a91e0d